v2/src/log_sink_etw.c

```diff
@@ -550,54 +549,70 @@ static void internal_emit_self_described_event(const char* event_name, uint16_t
-                internal_emit_self_described_event(event_name_info, sizeof(event_name_info), TRACE_LEVEL_INFORMATION, NULL, 0, __FILE__, __FUNCTION__, __LINE__, "ETW provider was registered succesfully (self test). Executable file full path name = %s", executable_full_path_name);
+            result = MU_FAILURE;
+        }
+        else
+        {
+            char* temp_executable_full_path_name;
+            const char* executable_full_path_name;
 
-                break;
+            if (_get_pgmptr(&temp_executable_full_path_name) != 0)
+            {
+                /* Codes_SRS_LOG_SINK_ETW_01_083: [ If _get_pgmptr fails, the executable shall be printed as UNKNOWN. ]*/
+                executable_full_path_name = "UNKNOWN";
             }
             else
             {
-                /* Codes_SRS_LOG_SINK_ETW_01_088: [ If TraceLoggingRegister fails, the state shall be switched to NOT_REGISTERED (1). ]*/
-                (void)printf("ETW provider was NOT registered (register_result=0x%08x).\r\n", register_result);
-                (void)InterlockedExchange(&etw_provider_state, PROVIDER_STATE_NOT_REGISTERED);
+                executable_full_path_name = temp_executable_full_path_name;
             }
+
+            /* Codes_SRS_LOG_SINK_ETW_01_008: [ log_sink_etw.init shall emit a LOG_LEVEL_INFO event as a self test, printing the fact that the provider was registered and from which executable (as obtained by calling _get_pgmptr). ]*/
+            internal_emit_self_described_event(event_name_info, sizeof(event_name_info), TRACE_LEVEL_INFORMATION, NULL, 0, __FILE__, __FUNCTION__, __LINE__, "ETW provider was registered succesfully (self test). Executable file full path name = %s", executable_full_path_name);
+
+            log_sink_etw_state = LOG_SINK_ETW_STATE_INITIALIZED;
+
+            /* Codes_SRS_LOG_SINK_ETW_01_091: [ log_sink_etw.init shall succeed and return 0. ] */
+            result = 0;
         }
     }
+    else
+    {
+        /* Codes_SRS_LOG_SINK_ETW_01_092: [ If the module is already initialized, log_sink_etw.init shall fail and return a non-zero value. ]*/
+        (void)printf("log_sink_etw_init called in %" PRI_MU_ENUM "\r\n", MU_ENUM_VALUE(LOG_SINK_ETW_STATE, log_sink_etw_state));
+        result = MU_FAILURE;
+    }
+
+    return result;
+}
+
+static void log_sink_etw_deinit(void)
+{
+    switch (log_sink_etw_state)
+    {
+    default:
+    case LOG_SINK_ETW_STATE_NOT_INITIALIZED:
+        /* Codes_SRS_LOG_SINK_ETW_01_093: [ If the module is not initialized, log_sink_etw.deinit shall return. ]*/
+        (void)printf("log_sink_etw_deinit called in %" PRI_MU_ENUM "\r\n", MU_ENUM_VALUE(LOG_SINK_ETW_STATE, log_sink_etw_state));
+        break;
+    case LOG_SINK_ETW_STATE_INITIALIZED:
+        /* Codes_SRS_LOG_SINK_ETW_01_090: [ log_sink_etw.deinit shall return. ] */
+        TraceLoggingUnregister(log_sink_etw_provider);
```

Comment:
spec just says "shall return" but it seems like it calls TraceLoggingUnregister

Reply:
Fixed, ty